libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -187,14 +171,157 @@ Algorithm selectAlgorithm(TangentSpaceMeshInput *input) {
+
+    const float2* uvs = input->uvs;
+    const size_t uvstride = input->uvStride ? input->uvStride : sizeof(float2);
+
+    const bool isTriangle16 = input->triangles16 != nullptr;
+    const uint8_t* triangles = isTriangle16 ? (const uint8_t*) input->triangles16 :
+            (const uint8_t*) input->triangles32;
+
+    const size_t triangleCount = input->triangleCount;
+    const size_t tstride = isTriangle16 ? sizeof(ushort3) : sizeof(uint3);
+
+    const size_t outVertexCount = triangleCount * 3;
+    float3* outPositions = new float3[outVertexCount];
+    float2* outUvs = uvs ? new float2[outVertexCount] : nullptr;
+    quatf* quats = new quatf[outVertexCount];
+
+    const size_t outTriangleCount = triangleCount;
+    uint3* outTriangles = new uint3[outTriangleCount];
+
+    size_t vindex = 0;
+    for (size_t tindex = 0; tindex < triangleCount; ++tindex) {
+        uint3 tri = isTriangle16 ?
+                uint3(*(ushort3*)(pointerAdd(triangles, tindex, tstride))) :
+                *(uint3*)(pointerAdd(triangles, tindex, tstride));
+
+        const float3 pa = *pointerAdd(positions, tri.x, pstride);
+        const float3 pb = *pointerAdd(positions, tri.y, pstride);
+        const float3 pc = *pointerAdd(positions, tri.z, pstride);
+
+        uint32_t i0 = vindex++, i1 = vindex++, i2 = vindex++;
+        outTriangles[tindex] = uint3{i0, i1, i2};
+
+        outPositions[i0] = pa;
+        outPositions[i1] = pb;
+        outPositions[i2] = pc;
+
+        const float3 n = normalize(cross(pc - pb, pa - pb));
+        float3 t, b;
+        frisvadKernel(n, t, b);
+
+        const quatf tspace = mat3f::packTangentFrame({t, b, n});
+        quats[i0] = tspace;
+        quats[i1] = tspace;
+        quats[i2] = tspace;
+
+        if (outUvs) {
+            outUvs[i0] = *pointerAdd(uvs, tri.x, uvstride);
+            outUvs[i1] = *pointerAdd(uvs, tri.y, uvstride);
+            outUvs[i2] = *pointerAdd(uvs, tri.z, uvstride);
+        }
+    }
+
+    output->tangentSpace = quats;
+    output->positions = outPositions;
+    output->vertexCount = outVertexCount;
+    output->uvs = outUvs;
+    output->triangles32 = outTriangles;
+    output->triangleCount = outTriangleCount;
+}
+
+template<typename DataType, typename InputType>
+inline void cleanOutputPointer(DataType*& ptr, InputType inputPtr) noexcept {
+    if (ptr && ptr != (const DataType*) inputPtr) {
+        delete[] ptr;
+    }
+    ptr = nullptr;
+}
+
+template <typename DataType>
+inline void takeStride(DataType*& out, size_t stride) noexcept {
+    out = (DataType*) (((uint8_t*) out) + stride);
+}
+
+template <typename OutputType>
+using QuatConversionFunc = OutputType(*)(const quatf&);
+
+template<typename OutputType>
+inline void getQuatsImpl(OutputType* UTILS_RESTRICT out, const quatf* UTILS_RESTRICT tangentSpace,
+        size_t vertexCount, size_t stride, QuatConversionFunc<OutputType> conversion) noexcept {
```

Comment:
here also wondering if the conversion shouldn't be a template, so it's inlined in the tight loop.

Reply:
done